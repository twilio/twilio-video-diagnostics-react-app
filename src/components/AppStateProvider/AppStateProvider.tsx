import React, { createContext, useContext, useReducer, useEffect } from 'react';
import produce from 'immer';
import { VideoInputTest } from '@twilio/rtc-diagnostics';

export enum ActivePane {
  GetStarted,
  DeviceCheck,
  DeviceError,
  CameraTest,
  Connectivity,
  Quality,
  Results,
}

interface stateType {
  activePane: ActivePane;
  videoGranted: boolean;
  audioGranted: boolean;
  deviceError: null | Error;
  videoInputTestReport: null | VideoInputTest.Report;
}

type ACTIONTYPE =
  | { type: 'set-active-pane'; newActivePane: ActivePane }
  | { type: 'next-pane' }
  | { type: 'previous-pane' }
  | { type: 'set-devices'; devices: MediaDeviceInfo[] }
  | { type: 'set-device-error'; error: Error }
  | { type: 'set-video-test-report'; report: VideoInputTest.Report };

type AppStateContextType = {
  state: stateType;
  dispatch: React.Dispatch<ACTIONTYPE>;
};

export const initialState = {
  activePane: ActivePane.GetStarted,
  videoGranted: false,
  audioGranted: false,
  deviceError: null,
  videoInputTestReport: null,
};

export const AppStateContext = createContext<AppStateContextType>(null!);

export function useAppStateContext() {
  const context = useContext(AppStateContext);
  if (!context) {
    throw new Error('useAppStateContext must be used within an AppStateProvider');
  }
  return context;
}

/**
 * This reducer function helps with the logic used to decide the activePane,
 * as well as which panes are to be shown to the user, and when.
 */

export const appStateReducer = produce((draft: stateType, action: ACTIONTYPE) => {
  switch (action.type) {
    case 'set-active-pane':
      draft.activePane = action.newActivePane;
      break;

    case 'next-pane':
      switch (draft.activePane) {
        case ActivePane.GetStarted:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.CameraTest;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        case ActivePane.DeviceCheck:
          if (draft.deviceError) {
            draft.activePane = ActivePane.DeviceError;
          } else {
            draft.activePane = ActivePane.CameraTest;
          }
          break;
        default:
          draft.activePane++;
          break;
      }
      break;

    case 'previous-pane':
      switch (draft.activePane) {
        case ActivePane.CameraTest:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.GetStarted;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        default:
          draft.activePane--;
          break;
      }
      break;

    case 'set-devices':
      draft.audioGranted = action.devices.filter((d) => d.kind === 'audioinput').every((d) => d.label);
      draft.videoGranted = action.devices.filter((d) => d.kind === 'videoinput').every((d) => d.label);
      break;

    case 'set-device-error':
      draft.deviceError = action.error;
      draft.activePane = ActivePane.DeviceError;
      break;

    case 'set-video-test-report':
      draft.videoInputTestReport = action.report;
  }
});

export const AppStateProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(appStateReducer, initialState);

  useEffect(() => {
    navigator.mediaDevices.enumerateDevices().then((devices) => dispatch({ type: 'set-devices', devices }));
  }, []);

  return <AppStateContext.Provider value={{ state, dispatch }}>{children}</AppStateContext.Provider>;
};
