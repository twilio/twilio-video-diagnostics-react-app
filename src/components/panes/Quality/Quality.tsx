import { ExcellentQuality } from './ExcellentQuality/ExcellentQuality';
import { PoorQuality } from './PoorQuality/PoorQuality';
import { getQualityScore } from './getQualityScore/getQualityScore';
import { useAppStateContext } from '../../AppStateProvider/AppStateProvider';
import { QualityModal } from './QualityModal/QualityModal';
import { useState } from 'react';

export enum QualityScore {
  Poor,
  Suboptimal,
  Good,
  Excellent,
}

export function Quality() {
  const { state } = useAppStateContext();
  const [isModalOpen, setIsModalOpen] = useState(false);
  const { totalQualityScore, latency, jitter, packetLoss, bitrate } = getQualityScore(
    state.preflightTest.report,
    state.bitrateTest.report
  );

  return (
    <>
      <QualityModal
        isModalOpen={isModalOpen}
        setIsModalOpen={setIsModalOpen}
        jitter={jitter}
        latency={latency}
        packetLoss={packetLoss}
        bitrate={bitrate}
      />
      {totalQualityScore === QualityScore.Good || totalQualityScore === QualityScore.Excellent ? (
        <ExcellentQuality quality={totalQualityScore} openModal={() => setIsModalOpen(true)} />
      ) : (
        <PoorQuality quality={totalQualityScore} openModal={() => setIsModalOpen(true)} />
      )}
    </>
  );
}
