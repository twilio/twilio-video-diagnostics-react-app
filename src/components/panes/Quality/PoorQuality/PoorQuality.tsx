import { Container, Grid, Typography, Button, Paper, makeStyles } from '@material-ui/core';
import { ErrorIcon } from '../../../../icons/ErrorIcon';
import { QualityScore } from '../Quality';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';
import { ViewIcon } from '../../../../icons/ViewIcon';
import { WarningIcon } from '../../../../icons/WarningIcon';

const useStyles = makeStyles({
  modal: {
    width: '6000px',
  },
  paper: {
    padding: '1.5em',
    borderRadius: '8px',
    '& li': {
      '&:not(:last-child)': {
        marginBottom: '1.5em',
      },
    },
  },
  paperContainer: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  errorIcon: {
    position: 'absolute',
    right: 'calc(100% + 18px)',
  },
  viewButton: {
    marginTop: '2em',
    '& svg': {
      position: 'relative',
      left: '-5px',
    },
  },
});

interface PoorQualityProps {
  quality: QualityScore | undefined;
  openModal: () => void;
}

export function PoorQuality({ quality, openModal }: PoorQualityProps) {
  const { nextPane } = useAppStateContext();
  const classes = useStyles();

  return (
    <Container>
      <Grid container alignItems="center" justifyContent="space-between">
        <Grid item md={5}>
          <Typography variant="h1" gutterBottom style={{ position: 'relative' }}>
            <div className={classes.errorIcon}>
              {quality === QualityScore.Suboptimal ? <WarningIcon /> : <ErrorIcon />}
            </div>
            Quality {quality === QualityScore.Suboptimal ? 'is okay' : 'issues'}
          </Typography>

          <Typography variant="body1" gutterBottom>
            This the last step! Your expected audio and video quality is{' '}
            <strong>{quality === QualityScore.Suboptimal ? 'suboptimal' : 'degraded'} </strong>
            and overall performance {quality === QualityScore.Suboptimal ? 'could be better' : 'will be poor'}.
          </Typography>

          <Typography variant="body1" gutterBottom>
            <strong>Did you try all of the tips to improve quality?</strong>
          </Typography>

          <Button variant="contained" color="primary" onClick={nextPane} style={{ marginRight: '1.5em' }}>
            Yes
          </Button>
          <Button color="primary" onClick={nextPane}>
            Skip for now
          </Button>
        </Grid>

        <Grid item md={5} className={classes.paperContainer}>
          <Paper className={classes.paper}>
            <Typography variant="body1">
              <strong>Tips to improve quality and performance:</strong>
            </Typography>

            <ul>
              <li>
                <Typography variant="body1">
                  Close unused tabs and applications on your computer/device when using video.
                </Typography>
              </li>
              <li>
                <Typography variant="body1">Move closer to your router.</Typography>
              </li>
              <li>
                <Typography variant="body1">Turn off or disconnect from VPN.</Typography>
              </li>
              <li>
                <Typography variant="body1">Restart your computer or device.</Typography>
              </li>
            </ul>
          </Paper>

          <Button variant="outlined" onClick={openModal} className={classes.viewButton}>
            <ViewIcon />
            View detailed quality information
          </Button>
        </Grid>
      </Grid>
    </Container>
  );
}
