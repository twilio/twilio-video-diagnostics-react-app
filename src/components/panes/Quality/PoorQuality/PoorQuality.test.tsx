import { shallow } from 'enzyme';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';
import { QualityScore } from '../Quality';
import { PoorQuality } from './PoorQuality';

jest.mock('../../../AppStateProvider/AppStateProvider');

const mockUseAppStateContext = useAppStateContext as jest.Mock<any>;

mockUseAppStateContext.mockImplementation(() => ({ nextPane: jest.fn() }));

const mockSetIsModalOpen = jest.fn();

describe('the PoorQuality component', () => {
  it('should render correctly when quality is "poor"', () => {
    const wrapper = shallow(<PoorQuality quality={QualityScore.Poor} openModal={mockSetIsModalOpen} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when quality is "suboptimal"', () => {
    const wrapper = shallow(<PoorQuality quality={QualityScore.Suboptimal} openModal={mockSetIsModalOpen} />);
    expect(wrapper).toMatchSnapshot();
  });
});
