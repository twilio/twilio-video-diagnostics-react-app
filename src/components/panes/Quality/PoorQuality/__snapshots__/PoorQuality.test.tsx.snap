// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the PoorQuality component should render correctly when quality is "poor" 1`] = `
<WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    justifyContent="space-between"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      md={5}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        style={
          Object {
            "position": "relative",
          }
        }
        variant="h1"
      >
        <div
          className="makeStyles-errorIcon-4"
        >
          <ErrorIcon />
        </div>
        Quality 
        issues
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        This the last step! Your expected audio and video quality is
         
        <strong>
          degraded
           
        </strong>
        and overall performance 
        will be poor
        .
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        <strong>
          Did you try all of the tips to improve quality?
        </strong>
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[MockFunction]}
        style={
          Object {
            "marginRight": "1.5em",
          }
        }
        variant="contained"
      >
        Yes
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[MockFunction]}
      >
        Skip for now
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-paperContainer-3"
      item={true}
      md={5}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <WithStyles(ForwardRef(Typography))
          variant="body1"
        >
          <strong>
            Tips to improve quality and performance:
          </strong>
        </WithStyles(ForwardRef(Typography))>
        <ul>
          <li>
            <WithStyles(ForwardRef(Typography))
              variant="body1"
            >
              Close unused tabs and applications on your computer/device when using video.
            </WithStyles(ForwardRef(Typography))>
          </li>
          <li>
            <WithStyles(ForwardRef(Typography))
              variant="body1"
            >
              Move closer to your router.
            </WithStyles(ForwardRef(Typography))>
          </li>
          <li>
            <WithStyles(ForwardRef(Typography))
              variant="body1"
            >
              Turn off or disconnect from VPN.
            </WithStyles(ForwardRef(Typography))>
          </li>
          <li>
            <WithStyles(ForwardRef(Typography))
              variant="body1"
            >
              Restart your computer or device.
            </WithStyles(ForwardRef(Typography))>
          </li>
        </ul>
      </WithStyles(ForwardRef(Paper))>
      <WithStyles(ForwardRef(Button))
        className="makeStyles-viewButton-5"
        onClick={[MockFunction]}
        variant="outlined"
      >
        <ViewIcon />
        View detailed quality information
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;

exports[`the PoorQuality component should render correctly when quality is "suboptimal" 1`] = `
<WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    justifyContent="space-between"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      md={5}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        style={
          Object {
            "position": "relative",
          }
        }
        variant="h1"
      >
        <div
          className="makeStyles-errorIcon-4"
        >
          <ErrorIcon />
        </div>
        Quality 
        is okay
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        This the last step! Your expected audio and video quality is
         
        <strong>
          suboptimal
           
        </strong>
        and overall performance 
        could be better
        .
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        <strong>
          Did you try all of the tips to improve quality?
        </strong>
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[MockFunction]}
        style={
          Object {
            "marginRight": "1.5em",
          }
        }
        variant="contained"
      >
        Yes
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[MockFunction]}
      >
        Skip for now
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-paperContainer-3"
      item={true}
      md={5}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-paper-2"
      >
        <WithStyles(ForwardRef(Typography))
          variant="body1"
        >
          <strong>
            Tips to improve quality and performance:
          </strong>
        </WithStyles(ForwardRef(Typography))>
        <ul>
          <li>
            <WithStyles(ForwardRef(Typography))
              variant="body1"
            >
              Close unused tabs and applications on your computer/device when using video.
            </WithStyles(ForwardRef(Typography))>
          </li>
          <li>
            <WithStyles(ForwardRef(Typography))
              variant="body1"
            >
              Move closer to your router.
            </WithStyles(ForwardRef(Typography))>
          </li>
          <li>
            <WithStyles(ForwardRef(Typography))
              variant="body1"
            >
              Turn off or disconnect from VPN.
            </WithStyles(ForwardRef(Typography))>
          </li>
          <li>
            <WithStyles(ForwardRef(Typography))
              variant="body1"
            >
              Restart your computer or device.
            </WithStyles(ForwardRef(Typography))>
          </li>
        </ul>
      </WithStyles(ForwardRef(Paper))>
      <WithStyles(ForwardRef(Button))
        className="makeStyles-viewButton-5"
        onClick={[MockFunction]}
        variant="outlined"
      >
        <ViewIcon />
        View detailed quality information
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;
